### ISSUE TITLE
Signal that is processed by workflow synchronously

### ISSUE THREAD

**Issue by mfateev** [uber/cadence#538]
The use case is to reject Signal if it cannot be processed by the workflow code for whatever reason. For example if Signal is created from a Cherami message handler nacking it would put it into Cherami DLQ. 
We could introduce new concept which is similar to Query but mutates the workflow state and can return result instead of Signal to cover this use case.

**Comment by samarabbas** [comment]
If we implement this feature, then we can rely on DLQ support from other technologies like Kafka or Cherami.  The idea would be to send signal in response to Kafka/Cherami message and then synchronously (could be long poll) to wait for successful processing of signal.  If the signal fails multiple times then we nack the original message and let it end up in the DLQ exposed by the messaging system.
Here are few other things discussed:
1) Let the current signal semantics function as is and create new event/api for request/reply signals
2) We need to think through reply could be long running and we need API to long for response for signal
3) Do not write signal into history until successfully processed.
4) We might need to dedupe signals and need a signal cache for deduping for this feature

**Comment by Groxx** [comment]
Personally, I'd like to be able to use this to e.g. tell a workflow to refresh its config, or retry the current operation, or that kind of thing.  We could use it to support a web UI that could submit a request *and tell you if it worked*.  Currently we'd need to build this kind of thing via a signal + polling (probably a query handler) until it's complete.

**Comment by bayesianmind** [comment]
I've been prototyping this. The implementation I'm working on is external to cadence but could be merged in. The api looks like this:
Callee workflow exposes functionality like so:
```
err := workflow.SetCallHandler(ctx, "MutateStateA", func(dataStruct ProjectState) (ReplyData, error) {
    // do something
    return data, nil
})
// err is thrown just like SetQueryHandler (invalid handler name, bad signature for func)
```

In caller workflow (caller could be non-workflow callers with server support):
```
// timeout for call can be customized by setting opts on ctx just like StartActivityOptions
fut := workflow.CallExternalWorkflow(ctx, "destWF", "", "MutateStateA", ProjectState{})
var reply ReplyData
err := fut.Get(ctx, &reply) // blocks until Call timeout or call finishes
// err is either cadence error or the error returned from call handler
```

Under the hood in my prototype this is done like so:
- Caller registers a signal handler for replies
- Caller sends info about itself along with the serialized call to callee. Assigns temp ID to a Future and returns the Future. If timeout is set, schedule delayed timeout completion.
- Callee WF runs the call handler and returns data by signaling back caller
- Caller receives signal, looks up the correct future, completes that future with the data from the signal if it has not timed out.

**Comment by mfateev** [comment]
@bayesianmind, your prototype makes sense for the workflow to workflow communication. But it is not solving the original use case of an external processes invoking the request/reply operation on a workflow and if this operation is not applicable not recording it into the history. Also, ideally the external client should be able to use a long poll to wait for a request completion the same way it is already possible to wait for a workflow completion.

**Comment by bayesianmind** [comment]
Agreed, I'm going to keep this to workflow to workflow for now. Doing calls to workflows from outside the workflow will require server support as far as I can tell. I'll decouple that from this issue since this prototype is insufficient for the original usecase.
